File: source/com/barrybecker4/simulation/liquid1/model/Particles.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.liquid1.model;

import java.util.HashSet;
import java.util.Random;

/**
 *  A set of particles in a grid. These particles represent the fluid.
 *
 *  @author Barry Becker
 */
public class Particles extends HashSet<Particle> {

    /** ensure that the runs are the same  */
    private static final Random RANDOM = new Random(1);

    /**
     * @param x cell location x
     * @param y cell location y
     * @param numParticles  number of particles to add.
     * @param grid  some grid to add to.
     */
    public void addRandomParticles( double x, double y, int numParticles, Grid grid)  {

        for ( int i = 0; i < numParticles; i++ ) {
            addParticle(x + RANDOM.nextDouble(), y + RANDOM.nextDouble(), grid);
        }
    }

    public void addParticle( double x, double y, Grid grid) {

        Cell cell = grid.getCell((int)x, (int)y);
        Particle p = new Particle( x, y, cell);
        this.add(p);
        cell.incParticles();
    }

}

File: scala-source/com/barrybecker4/simulation/liquid/model/Particles.scala
/*
 * // Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
 */

package com.barrybecker4.simulation.liquid.model

import scala.collection.mutable
import scala.util.Random


/**
  * A set of particles in a grid. These particles represent the fluid.
  * @author Barry Becker
  */
object Particles {
  /** Ensure that the runs are the same  */
  private val RANDOM = new Random(1)
}

class Particles extends mutable.HashSet[Particle] {

  /**
    * @param x            cell location x
    * @param y            cell location y
    * @param numParticles number of particles to add.
    * @param grid         some grid to add to.
    */
  private[model] def addRandomParticles(x: Double, y: Double, numParticles: Int, grid: Grid) = {
    for (i <- 0 until numParticles)
      addParticle(x + Particles.RANDOM.nextDouble, y + Particles.RANDOM.nextDouble, grid)
  }

  def addParticle(x: Double, y: Double, grid: Grid): Unit = {
    val cell = grid.getCell(x.toInt, y.toInt)
    val p = new Particle(x, y, cell)
    this += p
    cell.incParticles()
  }
}
